mlir/lib/Linker/IRMover.cpp

```diff
@@ -7,7 +7,354 @@
+
+  DenseSet<Operation *> valuesToLink;
+  std::vector<Operation *> worklist;
+  // Replace-all-uses-with worklist
+  std::vector<std::pair<Operation *, Operation *>> rauwWorklist;
+
+  void maybeAdd(Operation *val) {
+    if (valuesToLink.insert(val).second)
+      worklist.push_back(val);
+  }
+
+  std::optional<Error> foundError;
+  void setError(Error e) {
+    if (e)
+      foundError = std::move(e);
+  }
+
+  Error linkFunctionBody(Operation *dst, FunctionLinkageOpInterface src);
+  Error linkGlobalValueBody(Operation *dst, Operation *src);
+  bool shouldLink(Operation *dst, Operation *src);
+  Operation *copyGlobalVariableProto(Operation *src, bool forDefinition);
+  Operation *copyFunctionProto(Operation *src);
+  Operation *copyGlobalValueProto(Operation *src, bool forIndirectSymbol);
+  Operation *linkGlobalValueProto(Operation *src, bool forIndirectSymbol);
+  void flushRAUWorklist();
+
+  // TODO: Consider sharing with linker
+  std::optional<StringRef> getSymbol(Operation *op) {
+    if (auto symbolOp = dyn_cast<SymbolOpInterface>(op)) {
+      return {symbolOp.getName()};
+    }
+    return {};
+  }
+
+  // TODO: Probably rename to something like getLinkedToOperation
+  // TODO: COnsider if it should be shared with the linker instead of copying
+  // it.
+  Operation *getLinkedToGlobal(Operation *srcOp) {
+
+    // If the source has no name it can't link.  If it has local linkage,
+    // there is no name match-up going on.
+    auto SrcSymOpt = getSymbol(srcOp);
+    if (!SrcSymOpt)
+      return nullptr;
+
+    auto SrcSym = *SrcSymOpt;
+
+    if (auto srcGVL = dyn_cast<GlobalValueLinkageOpInterface>(srcOp)) {
+      if (srcGVL.hasLocalLinkage())
+        return nullptr;
+    }
+
+    // Is there a match in the destination symbol table?
+    // TODO: Should the SymbolTable be a member instead?
+    SymbolTable syms(composite);
+    Operation *DstOp = syms.lookup(SrcSym);
+
+    if (!DstOp)
+      return nullptr;
+
+    // If the dst-operation has local linkage, it shouldn't be linked
+    if (auto dstGVL = dyn_cast<GlobalValueLinkageOpInterface>(DstOp)) {
+      if (dstGVL.hasLocalLinkage())
+        return nullptr;
+
+      // TODO: Do a check on instrinsic function mismatch?
+      return dstGVL;
+    }
+
+    return nullptr;
+  }
+
+public:
+  MLIRLinker(Operation *composite, OwningOpRef<Operation *> srcOp,
+             ArrayRef<Operation *> valuesToLink)
+      : composite{composite}, src{std::move(srcOp)} {
+    for (Operation *op : valuesToLink)
+      maybeAdd(op);
+  }
+  Error run();
+
+  Operation *materialize(Operation *op);
+};
+
+bool MLIRLinker::shouldLink(Operation *dst, Operation *src) {
+  auto srcGVL = dyn_cast<GlobalValueLinkageOpInterface>(src);
+  if (!srcGVL)
+    return false;
+
+  if (valuesToLink.count(src) || srcGVL.hasLocalLinkage()) {
+    return true;
+  }
+
+  if (dst) {
+    if (auto dstGVL = dyn_cast<GlobalValueLinkageOpInterface>(dst))
+      if (!dstGVL.isDeclarationForLinkage())
```

Comment:
```suggestion
      if (!(dgv.isDeclaration() || dgv.hasAvailableExternallyLinkage())
```

Reply:
`isDeclarationForLinkage ` is not `isDeclarationForLinker`

Reply:
Could you please clarify? I don't follow.